verifier/providerverifier.go

```diff
@@ -0,0 +1,305 @@
+	"github.com/openpubkey/openpubkey/gq"
+	"github.com/openpubkey/openpubkey/pktoken"
+	"github.com/openpubkey/openpubkey/util"
+	oidcclient "github.com/zitadel/oidc/v2/pkg/client"
+)
+
+type DefaultProviderVerifier struct {
+	issuer          string
+	commitmentClaim string
+	options         ProviderVerifierOpts
+}
+
+type ProviderVerifierOpts struct {
+	// If ClientID is specified, then verification will require that the ClientID
+	// be present in the audience ("aud") claim of the PK token payload
+	ClientID string
+	// Specifies whether to skip the Client ID check, defaults to false
+	SkipClientIDCheck bool
+	// Custom function for discovering public key of Provider
+	DiscoverPublicKey func(ctx context.Context, kid string, issuer string) (jwk.Key, error)
+	// Allows for successful verification of expired tokens
+	SkipExpirationCheck bool
+}
+
+// Creates a new ProviderVerifier with required fields
+//
+// issuer: Is the OpenID provider issuer as seen in ID token e.g. "https://accounts.google.com"
+// commitmentClaim: the ID token payload claim name where the cicHash was stored during issuance
+func NewProviderVerifier(issuer, commitmentClaim string, options ProviderVerifierOpts) *DefaultProviderVerifier {
+	v := &DefaultProviderVerifier{
+		issuer:          issuer,
+		commitmentClaim: commitmentClaim,
+		options:         options,
+	}
+
+	// If no custom DiscoverPublicKey function is set, set default
+	if v.options.DiscoverPublicKey == nil {
+		v.options.DiscoverPublicKey = DiscoverProviderPublicKey
+	}
+
+	return v
+}
+
+func (v *DefaultProviderVerifier) Issuer() string {
+	return v.issuer
+}
+
+func (v *DefaultProviderVerifier) VerifyProvider(ctx context.Context, pkt *pktoken.PKToken) error {
+	// Check whether Audience claim matches provided Client ID
+	// No error is thrown if option is set to skip client ID check
+	if err := verifyAudience(pkt, v.options.ClientID); err != nil && !v.options.SkipClientIDCheck {
+		return err
+	}
+
+	alg, ok := pkt.ProviderAlgorithm()
+	if !ok {
+		return fmt.Errorf("provider algorithm type missing")
+	}
+
+	switch alg {
+	case gq.GQ256:
+		if err := VerifyGQSig(ctx, pkt); err != nil {
+			return fmt.Errorf("error verifying OP GQ signature on PK Token: %w", err)
+		}
+	case jwa.RS256:
+		opToken, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		pubKey, err := v.ProviderPublicKey(ctx, opToken)
+		if err != nil {
+			return fmt.Errorf("failed to get OP public key: %w", err)
+		}
+
+		token, err := pkt.Compact(pkt.Op)
+		if err != nil {
+			return err
+		}
+
+		if _, err := jws.Verify(token, jws.WithKey(alg, pubKey)); err != nil {
+			return err
+		}
+	}
+
+	if err := v.verifyCommitment(pkt); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// This function takes in an OIDC Provider created ID token or GQ-signed modification of one and returns
+// the associated public key
+func (v *DefaultProviderVerifier) ProviderPublicKey(ctx context.Context, token []byte) (jwk.Key, error) {
```

Comment:
Given the name for this function was `PublicKey` and the fact that this is a Provider is implied by the fact that this is a `ProviderVerifier`, I don't see the benefit of renaming this.

Why change the returned type from `crypto.PublicKey` to `jwk.Key`. crypto.PublicKey is a standard part of the golang crypto library whereas changing this to `jwk.Key` makes our interface dependent on the jwx library. A dependency on the `jwk.Key` in the function signature means switching libraries under the hood is more diffcult.

A while ago you pushed for a change to use `crypto.signer` in all function signatures and that made our function signatures much nicer for the same reason.

I'm not advocating for changing functions that have jwx object types in the function arguments but we shouldn't make the problem worse.

Reply:
So, I found this to be a significantly important change. I do agree that relying on the jwx libraries in functional signatures is not ideal, and I'm happy to create our own object that alias's it. However, we should be capturing the entire key including any options it may contain and not just the public key. I know that we're not really doing anything with it now but it seems unfortunate short-term thinking to be creating a pattern that ignores part of the jwks.

Reply:
> However, we should be capturing the entire key including any options it may contain and not just the public key.

Is the main option you are concerned here alg?